### DISCUSSION TITLE
Why does ale sometimes omit id/seq from LSP/tsserver messages?

### DISCUSSION THREAD

**Discussion by benjaminjkraft** [w0rp/ale#4575] [Q&A - Ask for help with problems]
My question is about this bit of code (and the same bit a few lines above for tsserver):
https://github.com/dense-analysis/ale/blob/af42e0b510365bb137567336386262565150f26b/autoload/ale/lsp.vim#L140-L142
Why the `if`? It seems like both [LSP](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage) and [tsserver](https://github.com/microsoft/TypeScript/blob/defe33afa1f3bc0d7f8dfb14f754172d84a156c9/src/server/protocol.ts#L185) consider `id`/`seq` required for all messages. (The blame goes back to the original commits to add LSP, so it isn't very enlightening.)

As far as I know this doesn't break anything in ordinary usage. I'm running into it for the following (admittedly strange) reason: I'm trying to do some performance profiling/logging on tsserver, running a proxy process that logs how long various commands take. But for commands without a `seq`, it's hard to know what's a response to what. (Obviously some messages from the server aren't a response to anything -- that's fine -- but e.g. `geterr` does have some kind of response message that it's hard to track.) So I wonder: would it break anything to just always set seq/id?

**Comment by w0rp** [comment]
The original version of the LSP spec ALE implemented did not require `id` to be set for a `NotificationMessage`. There's probably some way we could check which version of the protocol the server is using and set `id` on notifications. For `tsserver`, there might be something similar we could do for `seq`.
